src/zmalloc.c

```diff
@@ -87,10 +88,36 @@ void zlibc_free(void *ptr) {
 #define dallocx(ptr, flags) je_dallocx(ptr, flags)
 #endif
 
-#define update_zmalloc_stat_alloc(__n) atomic_fetch_add_explicit(&used_memory, (__n), memory_order_relaxed)
-#define update_zmalloc_stat_free(__n) atomic_fetch_sub_explicit(&used_memory, (__n), memory_order_relaxed)
+#if __STDC_NO_THREADS__
+#define thread_local __thread
+#else
+#include <threads.h>
+#endif
+
+/* A thread-local storage which keep the current thread's index in the used_memory_thread array. */
+static thread_local int thread_index = -1;
+/* MAX_THREADS_NUM = IO_THREADS_MAX_NUM(128) + BIO threads(3) + main thread(1). */
+#define MAX_THREADS_NUM 132
+static unsigned long long used_memory_thread[MAX_THREADS_NUM];
```

Comment:
Does this need to be atomic? These can be accessed by the main thread and written by another thread simultaneously.

Reply:
@zuiderkwast The `total_active_threads` is atomic when assigned to `thread_index`, each thread's `thread_index` should be unique in global,  so I think its safe for main thread and another thread to access their own element in the array.

Reply:
I mean the main thread reads all of these per-thread used memory variables, possibly at the same time abother thread updates its used memory at the same time. On amd64 maybe a write to a 64-bit integer is always atomic, but it's not universally true. Main thread can see some bits written and other bits unwritten.

Or I have missed something...?

Reply:
@zuiderkwast Thanks for pointing this out. I did some research for non x86 platform (Mainly ARM platform).

I have done some research about ARM platform. For aligned fundamental data type, it is atomic to access. In the document of "Arm Architecture Reference Manual for A-profile architecture" it is called **single-copy atomic**.

But I also read a conner case for ARM v7 https://developer.arm.com/documentation/ddi0406/c/Application-Level-Architecture/Application-Level-Memory-Model/Memory-types-and-attributes-and-the-memory-order-model/Atomicity-in-the-ARM-architecture?lang=en#:~:text=Multi%2Dcopy%20atomicity.-,Single%2Dcopy%20atomicity,one%20of%20the%20write%20operations.

If you have any concern, we can enable this optimization for only x86 specific.

Reply:
@lipzhu we should use the native word, `size_t` and then we should get single-copy atomicity on all platforms.

```suggestion
static size_t used_memory_thread[MAX_THREADS_NUM];
```

Reply:
> then we should get single-copy atomicity on all platforms

There is no such guarantee in the C standard as fare as I'm aware. To convince me, you need to find a link supporting this claim.

If we're convinced that a word is lock-free, then we can as well use an atomic type for it, since that wouldn't cost anything extra on platforms where it's already atomic. We should use relaxed memory ordering of course.

There are macros to see if various types are lock free, even small types like char: https://en.cppreference.com/w/c/atomic/ATOMIC_LOCK_FREE_consts and this generic check: 
and https://en.cppreference.com/w/c/atomic/atomic_is_lock_free

Reply:
@zuiderkwast @PingXie I revisit the document of atomic operation for ARM architecture, per my understanding, to meet the requirement for single-copy atomicity, we can declare it with aligned. (Same with PowerPC)

![image](https://github.com/valkey-io/valkey/assets/698621/57ea3c4f-3036-413a-94e7-1bedfcbae70b)

```c 
static __attribute__((aligned(sizeof(size_t)))) size_t used_memory_thread[MAX_THREADS_NUM];
```

Ref1: https://developer.arm.com/documentation/ddi0487/latest/
Ref2: https://developer.arm.com/documentation/ddi0406/c/Application-Level-Architecture/Application-Level-Memory-Model/Memory-types-and-attributes-and-the-memory-order-model/Atomicity-in-the-ARM-architecture?lang=en
Ref3: https://devblogs.microsoft.com/oldnewthing/20180814-00/?p=99485#:~:text=On%20the%20PowerPC%20600%20series,a%20mix%20of%20the%20two.

Reply:
>There is no such guarantee in the C standard as far as I'm aware.

@zuiderkwast this is not a concern of the C language but about the CPU's memory model.

Reply:
OK, so it's the CPU, not the C standard. Still, the compiler generates code for the target CPU architecture. If a word is atomic on the target architecture, the compiler can treat `_Atomic size_t` just like `size_t` and the problem is solved. It would also know if it needs to be aligned or not. Why would making this atomic cost anything extra for platforms where a word is already atomic?

```C
static _Atomic size_t used_memory_thread[MAX_THREADS_NUM];
```

This would make this code safe on any platform, wouldn't it?

If we're compiling for an z/Architecture, VAX or some other exotic architecture, the compiler would know what needs to be done to get atomic semantics. Just testing for 2-3 of the most popular architectures is not convincing. If you look at config.h, you can see how many architectures we try to detect endianess for:

```
#if defined(__i386__) || defined(__x86_64__) || defined(__amd64__) || defined(vax) || defined(ns32000) ||              \
    defined(sun386) || defined(MIPSEL) || defined(_MIPSEL) || defined(BIT_ZERO_ON_RIGHT) || defined(__alpha__) ||      \
    defined(__alpha)
#define BYTE_ORDER LITTLE_ENDIAN
#endif

#if defined(sel) || defined(pyr) || defined(mc68000) || defined(sparc) || defined(is68k) || defined(tahoe) ||          \
    defined(ibm032) || defined(ibm370) || defined(MIPSEB) || defined(_MIPSEB) || defined(_IBMR2) || defined(DGUX) ||   \
    defined(apollo) || defined(__convex__) || defined(_CRAY) || defined(__hppa) || defined(__hp9000) ||                \
    defined(__hp9000s300) || defined(__hp9000s700) || defined(BIT_ZERO_ON_LEFT) || defined(m68k) ||                    \
    defined(__sparc) || (defined(__APPLE__) && defined(__POWERPC__))
#define BYTE_ORDER BIG_ENDIAN
```

Reply:
Ok, got your point, add `_Atomic` leave this complex to compiler. But I checked the instruction, it remains the `lock add`.

![image](https://github.com/valkey-io/valkey/assets/698621/0710f7f1-8695-4f0e-a0b9-a9cb0461a518)

Reply:
Interesting. I wonder why it adds lock instruction. Did you use aligned + atomic?

Maybe we can avoid atomic for platforms known to be safe (amd64, arm64, powerpc) and use atomic as a fallback for everything else?

Reply:
> Interesting. I wonder why it adds lock instruction. 
``` c
_Atomic int a;
a++; // equals atomic_fetch_add_explicit(&a, 1, memory_order_relaxed)
```
>Did you use aligned + atomic?

Yes, `lock add` remained.

> Maybe we can avoid atomic for platforms known to be safe (amd64, arm64, powerpc) and use atomic as a fallback for everything else?

+1.

Reply:
>got your point, add _Atomic leave this complex to compiler. But I checked the instruction, it remains the lock add.

This is my understanding of  the `atomic` behavior too. Thanks for being thorough on this, @lipzhu